src/server/transaction.cc

```diff
@@ -606,10 +610,10 @@ bool Transaction::RunInShard(EngineShard* shard, bool txq_ooo) {
 
   bool is_concluding = coordinator_state_ & COORD_CONCLUDING;
 
-  // If we're the head of tx queue (txq_ooo is false), we remove ourselves upon first invocation
+  // If we're allowed, we remove ourselves upon first invocation from the queue,
   // and successive hops are run by continuation_trans_ in engine shard.
-  // Otherwise we can remove ourselves only when we're concluding (so no more hops will follow).
-  if ((is_concluding || !txq_ooo) && sd.pq_pos != TxQueue::kEnd) {
+  // Otherwise we can remove ourselves only when we're concluding (so no more hops follow).
+  if (sd.pq_pos != TxQueue::kEnd && (is_concluding || allow_q_removal)) {
```

Comment:
The logic looks right. 
Minor comment is that I understand that we allow queue removal only if we are concluding or we are the tx is head , therefore I think it will be more clear to pass is_head instead of allow_q_removal, as this param has no meaning when you call run transaction which is not in txq.

Reply:
it's actually not correct. I tried first `sd.pq_pos  == txq()->Head()` without passing anything. but for multi-hop transactions that are OOO we do not set `continous_transaction_`, therefore we may break consistency guarantees if we remove them even if they are at head.

Reply:
I dont understand than.
If there is a multi hop transaction that is OOO and it was at head position we will call the RunInShard with allow_q_removal = true because it will run from the flow of engine_shard.cc:669 rigtht?

Reply:
no, we do not enter there if there is a continuous transaction defined.
but we can also call from ` if (trans && disarmed) {` check  (the last invocation)